### PR TITLE
[SPARK-30625][SQL] Support `escape` as third parameter of the `like` function

### PR DESCRIPTION
**Pull request by MaxGekk**
### What changes were proposed in this pull request?
In the PR, I propose to transform the `Like` expression to `TernaryExpression`, and add third parameter `escape`. So, the `like` function will have feature parity with `LIKE ... ESCAPE` syntax supported by https://github.com/apache/spark/commit/187f3c17733f94aa3372caca355ad18ec1198f2f.

### Why are the changes needed?
The `like` functions can be called with 2 or 3 parameters, and functionally equivalent to `LIKE` and `LIKE ... ESCAPE` SQL expressions.

### Does this PR introduce any user-facing change?
Yes, before `like` fails with the exception:
```sql
spark-sql> SELECT like('_Apache Spark_', '__%Spark__', '_');
Error in query: Invalid number of arguments for function like. Expected: 2; Found: 3; line 1 pos 7
```
After:
```sql
spark-sql> SELECT like('_Apache Spark_', '__%Spark__', '_');
true
```

### How was this patch tested?
- Add new example for the `like` function which is checked by `SQLQuerySuite`
- Run `RegexpExpressionsSuite` and `ExpressionParserSuite`.

### CODE REVIEW

**Review by dongjoon-hyun — COMMENTED** [review]
Ur, actually, I expected a smaller change.

**Comment by MaxGekk** [comment]
Actually I did bigger change but reverted my refactoring in `RLike`. The change could be smaller if I didn't rename expression parameters but I thought `case class Like(left: Expression, right: Expression, escapeChar: Char)` looks slightly confusing since it is not binary expression any more, and `right` is not most right parameter.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala [hunk 1]
```diff
@@ -107,46 +109,65 @@ abstract class StringRegexExpression extends BinaryExpression
...
+case class Like(str: Expression, pattern: Expression, escape: Expression)
+  extends TernaryExpression with StringRegexExpression {
 
-  def this(left: Expression, right: Expression) = this(left, right, '\\')
+  def this(str: Expression, pattern: Expression) = this(str, pattern, Literal("\\"))
+
+  override def inputTypes: Seq[DataType] = Seq(StringType, StringType, StringType)
+  override def children: Seq[Expression] = Seq(str, pattern, escape)
+
+  private lazy val escapeChar: Char = if (escape.foldable) {
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala]: We need `lazy` here? I personally think its better to check error conditions as soon as possible.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala [hunk 2]
```diff
@@ -107,46 +109,65 @@ abstract class StringRegexExpression extends BinaryExpression
...
+    escape.eval() match {
+      case s: UTF8String if s != null && s.numChars() == 1 => s.toString.charAt(0)
+      case s => throw new AnalysisException(
+        s"The 'escape' parameter must be a string literal of one char but it is $s.")
+    }
+  } else {
+    throw new AnalysisException("The 'escape' parameter must be a string literal.")
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala]: Can you add tests for this path and line 131 (error cases)?

**Comment by maropu** [comment]
The refactoring still looks fine to me.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala (hunk 1 above)
Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala]: When I remove `lazy`, I get the exception:
```
makeCopy, tree: str#13580 LIKE pattern#13581 ESCAPE '@'
org.apache.spark.sql.catalyst.errors.package$TreeNodeException: makeCopy, tree: str#13580 LIKE pattern#13581 ESCAPE '@'
	at org.apache.spark.sql.catalyst.errors.package$.attachTree(package.scala:56)
	at org.apache.spark.sql.catalyst.trees.TreeNode.makeCopy(TreeNode.scala:435)
	at org.apache.spark.sql.catalyst.trees.TreeNode.mapChildren(TreeNode.scala:408)
...
Caused by: org.apache.spark.sql.AnalysisException: The 'escape' parameter must be a string literal.;
	at org.apache.spark.sql.catalyst.expressions.Like.<init>(regexpExpressions.scala:135)
```
See `escape` is `PrettyAttribute`:
Screen Shot 2020-01-26 at 19 51 34

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala (hunk 2 above)
Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala]: Added

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala (hunk 1 above)
Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala]: I got the error when I run new test in `DataFrameFunctionsSuite`. PrettyAttribute is not foldable.

Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala]: Ur, ok. Thanks for the check. The current one looks ok to me.

**Review by maropu — APPROVED** [review]
cc: @dongjoon-hyun @HyukjinKwon

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, @MaxGekk and @maropu .
Merged to master.

**Comment by rednaxelafx** [comment]
Sorry for posting this review post-hoc, but I strongly believe this PR should be reverted and redesigned.

There are both design and implementation issues with this PR:

1. Design: in order to get to feature parity between the `LIKE` syntax and the `like` built-in function, the latter should only accept `Char` argument for the `escapeChar` parameter, instead of the full-blown `Expression` -- the `LIKE` syntax only accepts a literal anyway. c.f.
https://github.com/apache/spark/blob/a7451f44d234a946668deb99bed8fdbd4b8ebf8e/sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4#L748
By not changing the `escapeChar` parameter to an `Expression`, we can avoid most of the complexity (and bugs) in this PR.

2. Implementation. The shim of `escapeChar = ...` to adapt the `Expression` argument back to a `Char` is very bad. See this example:
```scala
Seq(("abc", "%bc", "/")).toDF("s", "p", "esc").selectExpr("like(s, p, esc)").explain(true)
```
and on a freshly built Apache Spark master, you'll get:
```
== Parsed Logical Plan ==
'Project [unresolvedalias('like('s, 'p, 'esc), Some(org.apache.spark.sql.Column$$Lambda$2247/782587714@5cf3924c))]
+- Project [_1#3 AS s#10, _2#4 AS p#11, _3#5 AS esc#12]
   +- LocalRelation [_1#3, _2#4, _3#5]

== Analyzed Logical Plan ==
s LIKE p: boolean
org.apache.spark.sql.AnalysisException: The 'escape' parameter must be a string literal.;
== Optimized Logical Plan ==
org.apache.spark.sql.AnalysisException: The 'escape' parameter must be a string literal.;
== Physical Plan ==
org.apache.spark.sql.AnalysisException: The 'escape' parameter must be a string literal.;
```

It doesn't look like supporting full-blown `Expression` for `escapeChar` is worth the cost here.

**Comment by gatorsmile** [comment]
Which systems support this syntax?

**Comment by dongjoon-hyun** [comment]
Thank you for pointing that, @rednaxelafx .

**Comment by dongjoon-hyun** [comment]
According to the above comment, @MaxGekk is trying to revert this first from 3.0.0 and will target 3.1.0 later.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala (hunk 2 above)
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala]: This kind of thing should be done in `checkInputDataTypes`.

**Pull request closed without merging**